next/src/hooks/useLoading.tsx

```diff
@@ -0,0 +1,25 @@
+import { useState, useEffect } from 'react'
+import type { NextRouter } from 'next/router'
+
+const useLoading = (router: NextRouter) => {
+  const [isLoading, setIsLoading] = useState(false)
+
+ useEffect(() => {
+    const handleStart = (url: string) => url !== router.asPath && setIsLoading(true)
+    const handleComplete = () => setIsLoading(false)
+
+    router.events.on('routeChangeStart', handleStart)
+    router.events.on('routeChangeComplete', handleComplete)
+    router.events.on('routeChangeError', handleComplete)
+
+    return () => {
+      router.events.off('routeChangeStart', handleStart)
+      router.events.off('routeChangeComplete', handleComplete)
+      router.events.off('routeChangeError', handleComplete)
+    }
+  })
```

Comment:
みたいな感じでも書けるかもしれないです🙌
`router.events.off` なんですが、コンポーネントがマウントされていない時に、サブスクライブをやめる感じみたいなので、すでにトップレベルで呼んでいたら常に ON にしておいても問題かもですかね。
https://nextjs.org/docs/api-reference/next/router#usage-6
```suggestion
const useLoading = (router: NextRouter) => {
  const [isLoading, setIsLoading] = useState(false)

  const handleStart = useCallback((url: string) => url !== router.asPath && setIsLoading(true), [router.asPath])
  const handleComplete = useCallback(() => setIsLoading(false), [])

  useEffect(() => {
    router.events.on('routeChangeStart', handleStart)
    router.events.on('routeChangeComplete', handleComplete)
    router.events.on('routeChangeError', handleComplete)
  }, [handleComplete, handleStart, router.events])

  return isLoading
}
```

Reply:
@otegami 

> router.events.off なんですが、コンポーネントがマウントされていない時に、サブスクライブをやめる感じ

つまり... コンポーネントがマウントされていないときにサブスクライブされていても問題なさそうだから、router.event.off は必要なさそう?という感じですかね。
というか、この hooks が呼ばれている時点で component がマウントされていて、component が unmount されるときに、この router のインスタンスも破棄されるので、unmount されているときのことは考えなくて良さそうという感じですかね?

Reply:
ですです！
ごめんなさい、日本語がおかしくなっていて読みづらかったですね...

---

蛇足
フロントエンドでは、画面全体をローディングにはせずに表示できる要素は先に表示して、遅延で一部をローディング画面にすることが多い気がしています

ただ、最近出てきた Remix では、その方針をガラッと変えていて @koheitakahashi の方針に近い感じがしています！
- https://remix.run/
- https://zenn.dev/kaa_a_zu/articles/fbd06ca2cc3b86

Reply:
> フロントエンドでは、画面全体をローディングにはせずに表示できる要素は先に表示して、遅延で一部をローディング画面にすることが多い気がしています

なるほど、そういうトレンドなんですね〜。勉強になります!!

remix ってフレームワークがあるんですね 👀 
SSG ができなかったりと、現在のトレンドの逆方向行っている感じがして挑戦的ですね 👀

Reply:
6d7aef53e3be555fb5ab99eec92309863b68e2c9 こちらで対応しました!!